### DISCUSSION TITLE
Error when launching ironbar: foreign toplevel manager not present

### DISCUSSION THREAD

**Discussion by thmasq** [JakeStanger/ironbar#402] [Q&A] [answered]
#### The problem:
When trying to launch the program, my config seems to be correctly parsing, but the program panics due to lacking a foreign toplevel manager. I tried running it on both Gnome and the Pinnacle WM, but I get the same error every time.

#### More details:
I compiled the latest stable release of Ironbar, 0.13 with the latest stable rust toolchain, 1.75. The build dependencies are properly installed, and the compiler outputs no errors.

#### The error message:
When I run the program with the configuration files properly set in the XDG config directory, I always get the same error message:
```2024-01-10T02:52:15.389347Z  INFO ironbar: 92: Ironbar version 0.13.0
2024-01-10T02:52:15.389360Z  INFO ironbar: 93: Starting application
2024-01-10T02:52:15.389797Z ERROR ironbar::logging: 47: The application panicked (crashed).
Message:  foreign toplevel manager is not available: NotPresent
Location: src/clients/wayland/client.rs:129

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
2024-01-10T02:52:15.431830Z ERROR ironbar::logging: 47: The application panicked (crashed).
Message:  Failed to send message to channel: SendError { .. }
Location: src/clients/wayland/client.rs:255
```

Which refers to this line:
```rust
// at src/clients/wayland/client.rs:129
let foreign_toplevel_manager_delegate = ToplevelManagerState::bind(&globals, &qh)
                .expect("foreign toplevel manager is not available");
```

Is there something I'm missing to properly run Ironbar?

**Comment by JakeStanger** [comment, marked as answer]
Hi, Ironbar only supports wlroots-based wayland compositors (such as sway and hyprland). This is because under the hood it relies on protocols specific to these, such as the 'foreign toplevel manager' you see in your error.

GNOME and Pinnacle WM are not wlroots based and do not implement the required protocols, so ironbar will never run in these environments unfortunately.

**Reply by thmasq** [reply]
Ah, I see. I didn't mean to clutter the discussion threads, but I was expecting Ironbar to work on Smithay because of the acknowledgement in the readme of this repo. Thanks for the great work you're putting on this project, and hopefully Smithay will implement the necessary protocols necessary for this to work one day.

**Reply by JakeStanger** [reply]
No worries. I think there's some misunderstanding as to what Smithay is/does. Smithay is effectively tooling for Rust for building a compositor, or building a client. It is down to the compositor to implement these protocols. 

Hopefully eventually some of the wlroots protocols will work their way into the main wayland protocols, at which point GNOME / Deepin / etc may implement them, but since GNOME seem reluctant even then to implement some (such as layer-shell, which is fundamental to software like ironbar, waybar, eww) I wouldn't get your hopes up.

**Reply by thmasq** [reply]
You see, that was another reason why I thought Ironbar would work on Pinnacle. Waybar worked without any problem (I haven't configured it yet, so it's outputting some errors related to the default config). The screenshot I'm sharing was taken on the winit mode because Pinnacle doesn't implement screencopy yet, but I can confirm it works on udev mode. I was just dreaming about an all rust desktop
![Screenshot from 2024-01-10 15-46-31](https://github.com/JakeStanger/ironbar/assets/78622691/18c46f21-dc78-4a47-898a-4be825805507)

**Reply by JakeStanger** [reply]
Waybar also depends on layer-shell, so Pinnacle must have support for that; KDE does too so you can use Waybar there. I know Waybar does not have a hard requirement on `wlr_foreign_toplevel` and has support for other protocols (I think there's a KDE launcher for example). You might find when you come to configure it though that there's no option for a `focused` or `launcher` type module.

Ironbar does have that as a hard requirement because it'd be effectively useless in other environments at the moment, and I don't currently have any plans to expand support because it's a *lot* of work.

**Reply by yavko** [reply]
What if you added a feature that disables all modules that depend on that protocol, that could allow partial support.

**Reply by thmasq** [reply]
I was reading about the internals, and yeah, that sounds like it could work. That protocol only seems to be necessary for getting a context menu of the apps I have open on each of my desktops, which is a feature I could live without. I already had to disable clipboard support on the feature flags of Ironbar to progress a little bit on the chain of errors I was getting

**Reply by thmasq** [reply]
It could also make sense to make wlr foreign toplevel manager a feature flag for use cases where extra privacy is desired. There's an interesting discussion on the mail lists for wayland [here](https://www.mail-archive.com/wayland-devel@lists.freedesktop.org/msg39978.html).
While their concern makes sense, Gnome implementing a javascript API that does the exact same thing doesn't really make sense in my head. I got all of this from this discussion at Stack Overflow that you can find [here](https://stackoverflow.com/a/64030239).

**Reply by JakeStanger** [reply]
> That protocol only seems to be necessary for getting a context menu of the apps I have open on each of my desktops

Yeah, it's fundamental to the `launcher` and `focused` modules specifically. Sticking these (and the relevant client code) behind feature flags is an option that could get the bar up and running in any compositor that supports layer-shell. I am open to that, as this is exactly what the `clipboard` feature flag does already (that also relies on a godmode wlroots protocol). 

I suppose there are a good few users who have bars which don't show running applications, so it would definitely be a good change. It is not something at the top of my list, but absolutely feel free to open an issue and even go as far as taking on the work if you're up for it. 

Additionally, if opening up to non-wlroots environments, the docs would need a brush up to make sure it is clear which modules will and will not work across environments.

> While their concern makes sense, Gmome implementing a javascript API that does the exact same thing doesn't really make sense in my head.

They have certainly made some strange and controversial decisions for Wayland I think it's fair to say. The years-long client-side vs server-side window decorations arguments are a personal favourite... so GNOME support is still out of the question for the foreseeable future  ;)

**Reply by thmasq** [reply]
I think I could work on that, yes. It would be a very interesting learning experience for me, and having Ironbar work on my all rust desktop would be amazing. For now, I'm going to just open an issue and read the relevant codebase to see if I'm up to the task, but I like being challenged anyway ;)

**Comment by yavko** [comment]
There is rust based compositor that implemented the protocols needed, so you could use it, or reimplement what it changes in pinnaclewm https://github.com/YaLTeR/niri/releases/tag/v0.1.1

**Reply by thmasq** [reply]
That's pretty cool indeed, and I have great hopes for Niri. It's a shame we don't have all the most important features for a usable rust compositor in the same package; Pinnacle implements xwayland support natively, but supports few protocols, and Niri has the most important protocols implemented, but no plans to implement xwayland support.
